### ISSUE TITLE
Rework of loadout deployment code

### ISSUE THREAD

**Issue by halgari** [Nexus-Mods/NexusMods.App#312]
Due to #299 #245 and others, we're going to need to do a rework of the deployment and analysis code for the engine. This will revolve around a doc that's still WIP, I'll update this ticket once that's done. 

Steps (each as a separate PR):

- [ ] Integrate the new Archive Format
  - [ ] Recompress archives during analysis
  - [ ] Modify deployment code to pull from the new archives
- [ ] Rework the ModFile class structure to use interfaces (trait-like system) instead of a deep class hierarchy
- [ ] Add support for generated mods (RedMod, FNIS, Nemesis, etc.), not just generated mod files.
- [ ] Implement a fingerprinting filter system, so files can express dependencies on mods, file hashes, ect. to trigger file creation
- [ ] Implement filters and fingerprinting on mod sort rules, and generated files
- [ ] Rework any changes in generated files

**Comment by Sewer56** [comment]
Revised text for ARD-0005 below (previously #241), since it would fall under this issue now.  

This includes some minor corrections (e.g. `XZ` is just `LZMA` so no need to mention it, more realistic compression ratio numbers for comparing compression algorithms, etc. and some additional information).

There's also an important implementation detail I'd like to mention, but I'm not quite sure where to include it:

- Some already compressed data uses simple LZ77-style compression algorithms data with no additional post processing.  (e.g. no Huffman Coding).  
- ZStandard will apply some post processing, like Huffman, further saving compression ratios.  
- Some sample numbers in [Benchmarks](https://nexus-mods.github.io/NexusMods.Archives.Nx/Benchmarks/#block-size-recompressed-files) 
- I believe this also applies to Skyrim BSAs, e.g. `Textures0.bsa` should see a 20% size reduction IIRC. 

It might be worth keeping compression enabled for certain known file/archive formats, especially considering the fact our extraction is always I/O bottlenecked; so even a small file size saving of 10% still means faster extraction.

-----------

# Recompress Downloads

## Context

Modding sites like Nexus use various archive formats to house mods. The most frequently used formats include 
`.zip`, `.7z`, and to a lesser extent, `.rar`. Among these, some utilize 
[SOLID compression](https://en.wikipedia.org/wiki/Solid_compression), where the raw data content of the archive is 
compressed as a single unit or into several units. 

While SOLID compression enhances the compression ratio, it can slow down the extraction of individual files because 
unrelated data also has to be decompressed.

## Problem Statement

There are instances where we need to extract individual files from an archive, such as accessing a `Mod Metadata` 
file (e.g., `config.json`) or previewing texture files (e.g., `texture.dds`). However, the large SOLID block sizes, 
particularly with `.7z` archives, which have a default size of 4GB on the Normal Preset, can make this process slow.

Moreover, not all these archive formats are optimized for speed. `.7z`, for example, uses the LZMA2 codec by default, 
which favors compression ratios over compression speeds. This can result in prolonged extraction times, hindering 
efficiency. Therefore, optimizing extraction speed could pave the way for potential feature enhancements.

## Additional Complexities

A further challenge arises with some formats, such as `.7z`, due to their intricate structures. 
These formats often carry an excess of metadata or possess interoperability issues. For instance, the `.7z` library, 
a COM solution, is exclusive to Windows and lacks a cross-platform, ***high-performance*** native library alternative.

The need to resort to external programs, in this case, is far from ideal given the associated startup lag and the
inability to extract data in memory. This leads to unnecessary IO operations, especially when only a few files need 
to be read from the archive. The entire archive has to be extracted, only for the extracted files to be discarded 
once they have served their purpose. 

Absolutely, here's a cleaner version of the provided text:

## Why We Need This?

- Allow users to preview files such as textures and meshes without decompressing the entire archive.  
- Allow for parallel decompression of multiple files from the same archive.  
- Allow for faster decompression of single large files.  
- Reduce IO waste by not decompressing files that are never used.  

Compression formats are numerous and all involve a lot of tradeoffs.  

Roughly speaking, there are two main categories, those optimized for size (compression ratio) and those 
optimized for performance.  

* `LZMA`, `LZMA2`, `bzip2`, `Gzip` are all optimized for size, they are all slow to decompress.  
* `zstd`, `lz4`, `Brotli`, `DEFLATE` are all optimized for performance, they are all very fast to decompress, 
and have configuration options for time spent compressing vs size.  

In general the ***amount of bytes saved*** between these compression algorithms is in the range of ~2x.  

For example given a data set of DDS BC7 textures:  
- `LZ4`: 0.88  
- `LZMA2`: 0.69  

i.e. While LZ4 might save 12% of the file size in the texture case, LZMA2 saved 31%. 
This can be a couple hundred MBs... 

But the data set must also be considered, consider a very long log file:  
- `LZ4`: 0.058  
- `LZMA2`: 0.029  

Here, the 3% additional saving isn't particularly significant; however, in terms of decompression speed, the difference 
is stark, with LZ4 outpacing LZMA2 by more than 10x on a single thread.  

There's also additional considerations. For example, LZ4 might decompress faster, but if most users' storage can't read 
anywhere near as fast, what do we do? In that case, a slightly slower compression algorithm with better ratios may
be preferable so more data can be decompressed.

Choosing the compression format isn't the focus of this ADR. That decision is addressed in a separate 
[archiving library](https://nexus-mods.github.io/NexusMods.Archives.Nx/), resulting from this ADR. 
It's essential to note, however, that most common compression formats trade off performance penalties for 
poor compression ratios due to their reliance on dated technology (e.g., `DEFLATE`).

The purpose of this section is to highlight the trade-offs we are making. What might be ideal for server storage or 
long-term archival may not necessarily be the best solution for modding.

## Considered Options

### Option 1: Keep archives decompressed
Keep files decompressed on disk. Most mod managers do this. However, this has the disadvantage of requiring
a lot of disk space. On NTFS filesystems the default block size is 4KiB, so smaller files will waste space up
to the next multiple of 4KiB. 

Deleting an archive requires us to find every file that was extracted from that archive and delete it.
This is a lot of IO work for archives that contain many small files.

Technical Note: In NTFS, files can sometimes be stored directly in MFT, avoiding the size penalty; however this is very
rare as it only applies to tiny files under ~900 bytes.

### Option 2: Keep files in their original format
This is the option implemented in the app at the time of writing, it has the advantage of not requiring any extra 
disk space, and using high compression ratio algorithms. It has the disadvantage of requiring us to decompress the 
entire archive to find a few files, and the decompression algorithms are often slow and not parallelizable.

### Option 3: Recompress archives

In this option we would recompress archives into a format that is more suitable for modding. We would implement this
at the end of the file analysis phase. When a mod is downloaded it is currently automatically extracted into a temporary
folder and then analyzed. Once the analysis is complete we delete the temporary folder. In this option we would instead
recompress the archive into a format that is more suitable for our use case.

## Decision Outcome

Option 3 was chosen due to the advantages it provides over the other options. It allows us to keep acceptable compression
ratios while also getting a significant boost to decompression performance and "just in time" file extraction.

### Consequences (Bad)

- For NMA, we can no longer hash an archive to uniquely identify a download. The file is now recompressed and hash will 
  be different. We will have to store the original hash of the archive in the database and use that to determine
  what files are already downloaded.

- Users will have to use the app (or other compatible software) to extract downloads. Not a major problem as we will 
  provide an API for this, and a CLI command, but something we'll have to keep in mind.

### Consequences (Good)

- We can now extract files from archives without decompressing the entire archive. This will allow us to preview files
  such as textures and meshes without having to decompress the entire archive.  

- Good, we can now extract files from archives in parallel if we use a chunked data format (detailed in spec).  

- Good, we can use whatever compression method we want.  

- Good, built correctly we can write alloc free compression methods that decompress from mmaped memory to mmaped memory and don't involve pushing data through internal buffers.  

- Good, we can still delete the mod with a single OS call, we don't have to find every file that was extracted from the 
  archive and delete it.  

## File Format

[This is documented at NexusMods.Archives.Nx](https://nexus-mods.github.io/NexusMods.Archives.Nx/Specification/Overview/).

**Comment by erri120** [comment]
Can this be closed as part of #334?

**Comment by Sewer56** [comment]
I think it's good to close, just need to commit the above text into the main repo at some point.

**Comment by halgari** [comment]
Done as part of #334
